inline/src/parser/mod.rs

```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
+    enclosed: bool,
+    merge_whitespace: bool,
+}
+
 /// Parser of Unimarkup inline formatting. Implemented as an [`Iterator`], yields one
 /// self-contained Unimarkup [`Inline`] with every iteration.
 ///
 /// [`Iterator`]: Iterator
 /// [`Inline`]: crate::Inline
 #[derive(Debug, Clone)]
-pub struct Parser {
+pub struct Parser<'tokens> {
     /// Iterator over [`Token`]s found in Unimarkup input.
     ///
     /// [`Token`]: crate::Token
-    iter: Tokens,
+    iter: Tokens<'tokens>,
 
     /// Storage of [`Token`] already yielded from [`TokenIterator`] but not consumed in current
     /// iteration of parsing.
     ///
     /// [`Token`]: crate::Token
     /// [`TokenIterator`]: crate::TokenIterator
     /// [`Inline`]: crate::Inline
-    token_cache: Option<Token>,
+    token_cache: Option<Token<'tokens>>,
 
     /// Storage of parsed [`Inline`]s that should be returned before parsing next [`Inline`].
     inline_cache: VecDeque<Inline>,
 }
 
-impl Parser {
+impl<'tokens> Parser<'tokens> {
     /// Returns the next [`Token`] either from [`Lexer`] directly or from internal token cache.
     ///
     /// [`Token`]: crate::Token
     /// [`Lexer`]: crate::Lexer
-    fn next_token(&mut self) -> Option<Token> {
+    fn next_token(&mut self) -> Option<Token<'tokens>> {
         if self.token_cache.is_some() {
             self.token_cache.take()
         } else {
             self.iter.next()
         }
     }
 
-    fn parse_inline(&mut self, token: Token) -> Inline {
-        // opening/closing of tokens is resolved at lexing stage
-        // at this point we can simply parse
-        let kind = token.kind;
-        let start = token.span.start;
-        let mut end = token.span.end;
+    fn parse_plain(&mut self, start_token: Token, cxt: PlainContext) -> Inline {
+        let kind = if start_token.is_plain_whitespace() {
+            TokenKind::Plain
+        } else {
+            start_token.kind
+        };
 
-        if kind != TokenKind::Plain && !token.opens() {
-            let (content, span) = token.into_inner();
-            let content = InlineContent::Plain(PlainContent::new(content, span));
-            return Inline::as_plain_or_eol(content, kind);
-        }
+        let (tkn_str, mut span) = start_token.parts();
 
-        let mut content: InlineContent<_, _> = NestedContent::default().into();
+        let mut content = if cxt.enclosed {
+            // skip first (the enclosing) token
+            String::new()
+        } else {
+            String::from(tkn_str)
+        };
 
-        if kind == TokenKind::Plain {
-            content.append(InlineContent::from(token));
+        while let Some(next_token) = self.next_token() {
+            let enclosed_and_closes = cxt.enclosed && next_token.closes(Some(&start_token));
+            let not_enclosed_and_interrupted = !cxt.enclosed && next_token.kind != kind;
+
+            if enclosed_and_closes {
+                span.end = next_token.span.end;
+                break;
```

Comment:
Because the `next_token` closes the given `Inline` and is not needed except for `Span` information (as seen in code). Basically, we've found the end, and we do not that end except for it to signal that it is the end (of some `Inline`).